zkevm-circuits/src/evm_circuit/execution/returndatacopy.rs

```diff
@@ -0,0 +1,399 @@
+        block: &Block<F>,
+        _tx: &Transaction,
+        _call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let [dest_offset, data_offset, size] =
+            [0, 1, 2].map(|i| block.rws[step.rw_indices[i as usize]].stack_value());
+
+        self.data_offset.assign(
+            region,
+            offset,
+            Some(
+                data_offset.to_le_bytes()[..N_BYTES_MEMORY_ADDRESS]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+
+        let [last_callee_id, return_data_offset, return_data_size] = [
+            (3, CallContextFieldTag::LastCalleeId),
+            (4, CallContextFieldTag::LastCalleeReturnDataOffset),
+            (5, CallContextFieldTag::LastCalleeReturnDataLength),
+        ]
+        .map(|(i, tag)| {
+            let rw = block.rws[step.rw_indices[i as usize]];
+            assert_eq!(rw.field_tag(), Some(tag as u64));
+            rw.call_context_value()
+        });
+        self.last_callee_id.assign(
+            region,
+            offset,
+            Value::known(
+                last_callee_id
+                    .to_scalar()
+                    .expect("unexpected U256 -> Scalar conversion failure"),
+            ),
+        )?;
+        self.return_data_offset.assign(
+            region,
+            offset,
+            Value::known(
+                return_data_offset
+                    .to_scalar()
+                    .expect("unexpected U256 -> Scalar conversion failure"),
+            ),
+        )?;
+        self.return_data_size.assign(
+            region,
+            offset,
+            Value::known(
+                return_data_size
+                    .to_scalar()
+                    .expect("unexpected U256 -> Scalar conversion failure"),
+            ),
+        )?;
+
+        // assign the destination memory offset.
+        let memory_address =
+            self.dst_memory_addr
+                .assign(region, offset, dest_offset, size, block.randomness)?;
+
+        // assign to gadgets handling memory expansion cost and copying cost.
+        let (_, memory_expansion_cost) = self.memory_expansion.assign(
+            region,
+            offset,
+            step.memory_word_size(),
+            [memory_address],
+        )?;
+        self.memory_copier_gas
+            .assign(region, offset, size.as_u64(), memory_expansion_cost)?;
+
+        // rw_counter increase from copy lookup is `length` memory read & writes
+        let copy_rwc_inc = size + size;
+        self.copy_rwc_inc.assign(
+            region,
+            offset,
+            Value::known(
+                copy_rwc_inc // 1 read & 1 write
+                    .to_scalar()
+                    .expect("unexpected U256 -> Scalar conversion failure"),
+            ),
+        )?;
+
+        let bytes_to_copy = return_data_size - (data_offset + size);
```

Comment:
`bytes_to_copy` a bit of strange name because this is not the number of bytes that are copied. Maybe better to not create a variable to store this value and just calculate it on the fly below.

Reply:
In my opinion a variable makes it more readable, I think we should say write a name that means what it actually is.

Reply:
OK, I will keep the variable if I find a proper name :)